go/private/go_repositories.bzl

```diff
@@ -134,22 +136,51 @@ filegroup(
   name = "go_include",
   srcs = [ "pkg/include" ],
 )
-"""
 
+go_root(
+  name = "go_root",
+  path = "{goroot}",
+)
+"""
 
 def _go_repository_select_impl(ctx):
-  os = ctx.os.name
-  # NOTE: This mapping cannot be table-driven to prevent
-  # Bazel from downloading the other archive.
-  if os == 'linux':
-    goroot = ctx.path(ctx.attr._linux).dirname
-  elif os == 'mac os x':
-    goroot = ctx.path(ctx.attr._darwin).dirname
+  rules_goroot = ctx.os.environ.get("RULES_GOROOT", None)
+
+  # 1. Configure the goroot path
+  if rules_goroot:
+    goroot = ctx.path(rules_goroot)
   else:
-    fail("unsupported operating system: " + os)
+    os_name = ctx.os.name
+    if os_name == 'linux':
+      goroot = ctx.path(ctx.attr._linux).dirname
+    elif os_name == 'mac os x':
+      goroot = ctx.path(ctx.attr._darwin).dirname
+    else:
+      fail("Unsupported operating system: " + os_name)
+
+  # 2. Create the symlinks and write the BUILD file.
+  gobin = goroot.get_child("bin")
+  gopkg = goroot.get_child("pkg")
+  gosrc = goroot.get_child("src")
+  ctx.symlink(gobin, "bin")
+  ctx.symlink(gopkg, "pkg")
+  ctx.symlink(gosrc, "src")
+
+  ctx.file("BUILD", GO_TOOLCHAIN_BUILD_FILE.format(
+    goroot = goroot,
+  ))
+
+  # 3. If the user has specified the goroot explicitly, confirm a
+  # functional installation.
+  if rules_goroot:
+    go = gobin.get_child("go")
+    result = ctx.execute([go, "env"])
+    if result.return_code:
+      fail("""
+Something's not right.  Are you sure '%s' points to a functional GOROOT?
```

Comment:
Any reason to make this multi-line?  It reads a little awkwardly and I would suggest putting back on a single line